Reduce unsafeness in WebCore/xml some more

annevk opened the pull request:
#### 6255629e282868120f6afbca98fe44699d866463
<pre>
Reduce unsafeness in WebCore/xml some more
<a href="https://bugs.webkit.org/show_bug.cgi?id=295484">https://bugs.webkit.org/show_bug.cgi?id=295484</a>

Reviewed by Chris Dumez.

Helps with <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

Canonical link: <a href="https://commits.webkit.org/297055@main">https://commits.webkit.org/297055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5009e924a538adc6322f70924b850a7a7489a5a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110452 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30111 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20544 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116477 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60706 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38698 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/83997 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113400 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/24575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/99457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64438 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/23939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/17596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60273 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/93960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/17652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119268 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37492 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27824 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37866 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/95725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/92788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/37787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/15522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17808 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37387 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42858 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37049 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38758 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/xml/XMLHttpRequest.h:
```diff
@@ -54,13 +55,15 @@ class XMLHttpRequestUpload;
 template<typename> class ExceptionOr;
 
-class XMLHttpRequest final : public ActiveDOMObject, public RefCounted<XMLHttpRequest>, private ThreadableLoaderClient, public XMLHttpRequestEventTarget {
+class XMLHttpRequest final : public ActiveDOMObject, public RefCountedAndCanMakeWeakPtr<XMLHttpRequest>, private ThreadableLoaderClient, public XMLHttpRequestEventTarget {
```
This change is not needed since we already subclass `CanMakeWeakPtr` from `ThreadableLoaderClient`

annevk replied:
Thanks for the review! The problem I kept running into is that it also inherits it from EventTarget, I think. And I don't know how to make it clear to the compiler to just pick one. I uploaded my attempt so you can see.

annevk replied:
Youenn helped me out a bit and I think it's good now.

cdumez commented on Source/WebCore/xml/XMLHttpRequest.h:
```diff
@@ -54,13 +55,15 @@ class XMLHttpRequestUpload;
     void ref() const final { RefCounted::ref(); }
     void deref() const final { RefCounted::deref(); }
 
+    USING_CAN_MAKE_WEAKPTR(CanMakeWeakPtr<XMLHttpRequest>);
```
Not needed. This should have been your hint that you are doing something odd.

annevk replied:
I have added this back, but now with `EventTarget` as argument.

cdumez commented on Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.h:
```diff
@@ -72,8 +72,10 @@ class XMLHttpRequestProgressEventThrottle : public CanMakeWeakPtr<XMLHttpRequest
+
     // Weak pointer to our XMLHttpRequest object as it is the one holding us.
-    XMLHttpRequest& m_target;
+    WeakPtr<XMLHttpRequest> m_target;
```
WeakRef since it cannot be null, based on previous code.

cdumez commented on Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.h:
```diff
@@ -72,8 +72,10 @@ class XMLHttpRequestProgressEventThrottle : public CanMakeWeakPtr<XMLHttpRequest
     void flushProgressEvent();
     void dispatchEventWhenPossible(Event&);
 
+    RefPtr<XMLHttpRequest> protectedTarget() { return m_target.get(); }
```
Should return a `Ref`.

cdumez commented on Source/WebCore/xml/XMLHttpRequestUpload.h:
```diff
@@ -52,9 +52,9 @@ class XMLHttpRequestUpload final : public XMLHttpRequestEventTarget {
+    ScriptExecutionContext* scriptExecutionContext() const final { return m_request->scriptExecutionContext(); }
 
-    XMLHttpRequest& m_request;
+    WeakPtr<XMLHttpRequest> m_request;
```
`WeakRef`

cdumez commented on Source/WebCore/xml/XPathValue.cpp:
```diff
@@ -117,7 +117,7 @@ String Value::toString() const
         if (m_data->nodeSet.isEmpty())
             return emptyString();
-        return stringValue(m_data->nodeSet.firstNode());
+        return stringValue(RefPtr { m_data->nodeSet.firstNode() }.get());
```
Would be nice to introduce a `protectedFirstNode()` for this one.

annevk replied:
As far as I can tell this is the only call to `firstNode()` that needs protecting and it seems unlikely we'll introduce new XPath code anytime soon. Happy to do it, but it seemed superfluous to me.

cdumez commented on Source/WebCore/xml/XMLHttpRequestProgressEventThrottle.cpp:
```diff
@@ -145,17 +149,22 @@ void XMLHttpRequestProgressEventThrottle::suspend()
     });
 }
 
+Ref<XMLHttpRequest> XMLHttpRequestProgressEventThrottle::protectedTarget()
```
Can be inlined in the header.

annevk committed:
Reduce unsafeness in WebCore/xml some more
https://bugs.webkit.org/show_bug.cgi?id=295484

Reviewed by Chris Dumez.

Helps with https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

Canonical link: https://commits.webkit.org/297055@main
